programs/margin-account/src/lib.rs

```diff
@@ -33,16 +37,95 @@ pub mod margin_account {
+        ctx: Context<OpenPositionAMM>,
+        amount_in: u64,
+        minimum_amount_out: u64,
+    ) -> ProgramResult {
+        let accounts = ctx.accounts.to_account_infos();
+
+        // create the desired swap amount and minimum amout of slippage the user is willing to sustain
+        let swap = spl_token_swap::instruction::Swap {
+            amount_in,
+            minimum_amount_out,
+        };
+
+        let instruction = &spl_token_swap::instruction::swap(
+            ctx.accounts.swap_program.key,
+            ctx.accounts.token_program.key,
+            ctx.accounts.swap_info.key,
+            ctx.accounts.swap_authority.key,
+            ctx.accounts.vault_signer.key,
+            ctx.accounts.loaned_vault.to_account_info().key,
+            ctx.accounts.swap_source.key,
+            ctx.accounts.swap_dest.key,
+            ctx.accounts.collateral_vault.to_account_info().key,
+            ctx.accounts.pool_mint.key,
+            ctx.accounts.pool_fee.key,
+            Some(ctx.accounts.host_fee.key),
+            swap,
+        )?;
+
+        let seeds = &[
+            ctx.accounts.margin_account.to_account_info().key.as_ref(),
+            &[ctx.accounts.margin_account.nonce],
+        ];
+        let signer = &[&seeds[..]];
+
+        invoke_signed(instruction, &accounts[1..], signer)?;
+
+        // Mark account as having an open trade
+        let margin_account = &mut ctx.accounts.margin_account;
+        margin_account.collateral_vault = *ctx.accounts.collateral_vault.to_account_info().key;
+
         Ok(())
     }
     /// Close an open leveraged position.
-    pub fn close_position(_ctx: Context<ClosePosition>) -> ProgramResult {
-        // TODO
+    pub fn close_position_amm(
+        ctx: Context<ClosePositionAMM>,
+        amount_in: u64,
+        minimum_amount_out: u64,
+    ) -> ProgramResult {
+        let accounts = ctx.accounts.to_account_infos();
+
+        let swap = spl_token_swap::instruction::Swap {
+            amount_in,
+            minimum_amount_out,
+        };
+
+        let instruction = &spl_token_swap::instruction::swap(
+            ctx.accounts.swap_program.key,
+            ctx.accounts.token_program.key,
+            ctx.accounts.swap_info.key,
+            ctx.accounts.swap_authority.key,
+            ctx.accounts.vault_signer.key,
+            ctx.accounts.collateral_vault.to_account_info().key,
+            ctx.accounts.swap_source.key,
+            ctx.accounts.swap_dest.key,
+            ctx.accounts.loaned_vault.to_account_info().key,
+            ctx.accounts.pool_mint.key,
+            ctx.accounts.pool_fee.key,
+            Some(ctx.accounts.host_fee.key),
+            swap,
+        )?;
+
+        let seeds = &[
+            ctx.accounts.margin_account.to_account_info().key.as_ref(),
+            &[ctx.accounts.margin_account.nonce],
+        ];
+        let signer = &[&seeds[..]];
+
+        invoke_signed(instruction, &accounts[1..], signer)?;
+
+        Ok(())
+    }
```

Comment:
I don't see why we need to wrap these, and these not just be CPI calls within open and close position. Seems weird to wrap these? I guess it's fine for now, but I can't see the necessity other than hitting the compute cap

Reply:
I think the only way to avoid this is to write a way to build the tx offline and the account just sign it. The part here are the funds are being sent to different places when opening and closing.

Reply:
nvm, was my misread